src/modules/knowledge_formalization/formalization.py

```diff
@@ -0,0 +1,517 @@
+            result_array_idx = eigenValueIdx
+            break
+    if result_array_idx == -1:
+        print("No EigenValue 1.0 in received eigenvalues. Exiting program...")
+        exit(3)
+    print("extracted eigenvectors successfully")
+
+    # remove eigen values from memory
+    del eigenvalues
+
+    # only keep real value
+    print("converting vectors to keep only real values")
+    resultArray = vectors.real
+
+    # remove eigen vectors from memory
+    del vectors
+
+    # normalize data from column with index result_array_idx because that column represents our results
+    print("normalizing data")
+    normalizedArray = normalize_data(resultArray.T[result_array_idx])
+    print("data normalized")
+
+    print("creating array of similar concepts")
+    similar_concepts = {}
+    for concept_id in range(len(normalizedArray.T)):
+        value = normalizedArray.T[concept_id]
+
+        # if ID is not the same as initial concept, extract the value
+        if concept_id not in initial_concepts:
+            similar_concepts[concept_id] = value
+
+    # sort descending - most similar concept is the highest
+    print("sorting array of similar concepts descdending by probability")
+    sorted_similar_concepts = sorted(similar_concepts.items(), key=operator.itemgetter(1), reverse=True)
+
+    # extract as many concepts as wanted by variable number_of_wanted_concepts
+    extracted_concepts = sorted_similar_concepts[:number_of_wanted_concepts]
+
+    # print result
+    for final_concept in extracted_concepts:
+        id = final_concept[0]
+        probability = final_concept[1]
+        old_id = new_old_index_mapping[id]
+        word = id_string_map[id]
+
+        print("Concept ID", id, "probability", probability, "word", word, "old ID", old_id)
+
+    exit(0)
+
+
+# method extracts concepts IDs and creates new dictionary from them, if dump doesn't exist yet. if dump exists,
+# it just reads it from a dump file
+def create_concept_ids(default_concept_file,
+                       new_old_id_mapping_dump_path,
+                       id_concept_mapping_json_dump_path,
+                       old_new_id_dict):
+
+    new_old_id_dict = {}
+    if len(old_new_id_dict) != 0:
+        # old_new_id_temp_dict contains the mapping old -> new ID for old concepts that contain transitions
+        # so far we have old -> new and new -> old ID dictionary
+        new_old_id_dict = {v: k for k, v in old_new_id_dict.items()}
+
+    new_id_to_concept_string_mapping = {}
+
+    if os.path.isfile(id_concept_mapping_json_dump_path) and os.path.isfile(new_old_id_mapping_dump_path):
+
+        # try opening concepts file dump and if it doesn't exist, try opening the file and construct dump from it
+        print("concept file json dumps exist")
+
+        print("opening file", new_old_id_mapping_dump_path)
+        with open(new_old_id_mapping_dump_path, 'r') as fp1:
+            new_old_dict_load = json.load(fp1)
+        fp1.close()
```

Comment:
You don't need to close this since you are using `with` construct that will automatically close the opened resource upon exiting the nested code block

Reply:
fixed